### PR TITLE
[RLlib; Offline RL] - Enable single-learner/multi-learner GPU training.

### LINKED ISSUE #50053: [RLlib 2.41.0] get_device only returns cpu this causes num_gpus_per_env_runner and num_gpus_per_learner not working
### What happened + What you expected to happen

The following code:
```
from ray.rllib.utils import try_import_torch

torch, _ = try_import_torch()

from ray.air._internal.torch_utils import get_devices

devices = get_devices()
print(devices)
```
returns "[device(type='cpu')]"
Even tough pytorch has full access to a cuda device.
`torch.device(0)`
is returning the cuda:0 device.

If I just replace the function to return the cuda device, the model is running perfectly on the gpu.

This function is used in multi_agent_env_runner and torch_learner causing the model to be transferred to the cpu.
https://github.com/ray-project/ray/blob/master/rllib/env/multi_agent_env_runner.py#L94
https://github.com/ray-project/ray/blob/master/rllib/core/learner/torch/torch_learner.py#L449

### Versions / Dependencies
It was running fine with ray 2.40.0. The issue is reproduceable with 2.41.0 running on linux and windows
Tested with Pytorch 2.3.1 and 2.51. Same issue.

### Reproduction script

```
from ray.rllib.utils import try_import_torch

torch, _ = try_import_torch()

from ray.air._internal.torch_utils import get_devices

devices = get_devices()
print(devices)
print("")
print(torch.device(0))
```

### Issue Severity

None

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

GPU training in Offline RL was not implemented, yet. This PR proposes multiple changes to enable GPU training in single- and multi-learner mode:
- Disables the `NumpyToTensor` connector in the learner connector pipeline in case of `num_gpus_per_learner>0`
- Changes the GPU assignment in the `TorchLearner` manages by `ray.rllib.utils.framework.get_devices` by adding a test for the `WORKER_MODE` - a local learner is not in `WORKER_MODE` and therefore will not get a GPU otherwise (note, this holds true for ALL RLlib ALGOS).
- Adds another test to the `Learner.update_from_batch_or_episodes` to test for `MultiAgentBatch`es that are not  holding tensors, yet - this is the offline local learner GPU case.
- Modifies the iteration loop in `Learner.update_from_iterator` to use 
      - An outer loop that takes care of iterator exhaustion in case  the desired number of updates per RLlib iteration are not run, yet.
      - Metrics for the `dataset_num_iters_per_learner` to get an eye on desired vs. realized updates per RLlib iteration.
- Removes the `learner` parameter from the `OfflinePreLearner` ctor as it is no longer needed.
- Instead adds a default `module_spec` and `module_state` to the `OfflinePreLearner` ctor and adds corresponding logic to `Algorithm.build` to provide these two arguments during initialization.
- Added functionality for single-learner setups to make use of `Learner.update_from_iterator` which enables faster updates because each iterator is run without interruption. This includes changes to `LearnerGroup._update` to handle cases with a `list` `batch` that contains a single `DataIterator`.
- Modified the `OfflineData.sample` calls in all Offline RL algos to use an iterator for updates as long as `dataset_num_iters_per_learner !=1`.
- Added single GPU tests for all Offline RL algorithms to `BUILD`.

## Related issue number

Closes #50053 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/BUILD [hunk 1]
```diff
@@ -325,6 +325,18 @@ py_test(
     ],
     args = ["--as-test", "--enable-new-api-stack"]
 )
+py_test(
```
Comment (sven1977) [inline rllib/BUILD:328]: Noice!!!

File: rllib/BUILD [hunk 2]
```diff
@@ -340,6 +352,19 @@ py_test(
     ],
     args = ["--as-test", "--enable-new-api-stack"]
 )
+# GPU training.
+py_test(
```
Comment (sven1977) [inline rllib/BUILD:356]: Noice!!!

File: rllib/BUILD [hunk 3]
```diff
@@ -525,6 +550,19 @@ py_test(
     ],
     args = ["--as-test", "--enable-new-api-stack"]
 )
+# GPU-training.
+py_test(
```
Comment (sven1977) [inline rllib/BUILD:554]: 💥

File: rllib/algorithms/algorithm.py [hunk 4]
```diff
@@ -856,13 +856,12 @@ def setup(self, config: AlgorithmConfig) -> None:
...
                     # updating during preprocessing.
                     self.offline_data.learner_handles = self.learner_group._workers
-                    # Provide the module_spec. Note, in the remote case this is needed
-                    # because the learner module cannot be copied, but must be built.
-                    self.offline_data.module_spec = module_spec
                 # Otherwise we can simply pass in the local learner.
                 else:
                     self.offline_data.learner_handles = [self.learner_group._learner]
-
+                # Provide the module_spec. Note, in the remote case this is needed
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:862]: nit: Can we add a TODO here to replace these set-some-object's-attributes-directly? We should find some solution for this in the future, an API, or setting these in the OfflineData constructor?

File: rllib/algorithms/cql/cql.py [hunk 5]
```diff
@@ -213,6 +213,10 @@ def build_learner_connector(
             AddNextObservationsFromEpisodesToTrainBatch(),
         )
 
+        # If training on GPU, do not convert batches to tensors.
+        if self.num_gpus_per_learner > 0:
```
Comment (sven1977) [inline rllib/algorithms/cql/cql.py:217]: nice!

File: rllib/algorithms/cql/cql.py [hunk 6]
```diff
@@ -307,7 +311,9 @@ def training_step(self) -> None:
             batch_or_iterator = self.offline_data.sample(
                 num_samples=self.config.train_batch_size_per_learner,
                 num_shards=self.config.num_learners,
-                return_iterator=self.config.num_learners > 1,
+                return_iterator=self.config.dataset_num_iters_per_learner > 1
```
Comment (sven1977) [inline rllib/algorithms/cql/cql.py:314]: nit: Explain this logic in a 1-line comment

File: rllib/algorithms/marwil/marwil.py [hunk 7]
```diff
@@ -461,7 +465,9 @@ class (multi-/single-learner setup) and evaluation on
             batch_or_iterator = self.offline_data.sample(
                 num_samples=self.config.train_batch_size_per_learner,
                 num_shards=self.config.num_learners,
-                return_iterator=self.config.num_learners > 1,
+                return_iterator=self.config.dataset_num_iters_per_learner > 1
```
Comment (sven1977) [inline rllib/algorithms/marwil/marwil.py:468]: same: add 1-line comment ... why?

File: rllib/core/learner/learner.py [hunk 8]
```diff
@@ -55,6 +55,8 @@
 from ray.rllib.utils.framework import try_import_tf, try_import_torch
 from ray.rllib.utils.metrics import (
     ALL_MODULES,
+    DATASET_NUM_ITERS_PER_LEARNER_TRAINED,
```
Comment (sven1977) [inline rllib/core/learner/learner.py:58]: 2 dumb questions:

* Will we put all OfflineData results into a subkey: `offline_data`? Similar to `env_runners` or `learners`?
* If yes, should we rename these into `NUM_DATASET_ITERS_...`
* I feel like we shouldn't do this per Learner, but rather per main Algorithm iteration. Meaning, we log/create the metric as `reduce=sum, clear_on_reduce=True` (`False` for the lifetime one) and then report it as a sum across all Learners. This would then be analogous to the online stats, e.g. `NUM_ENV_STEPS_TRAINED` and `NUM_ENV_STEPS_TRAINED_LIFETIME`. What do you think?

File: rllib/core/learner/learner.py [hunk 9]
```diff
@@ -1180,6 +1184,18 @@ def _finalize_fn(batch: Dict[str, numpy.ndarray]) -> Dict[str, Any]:
                 value=loss,
                 window=1,
             )
+        # Record the number of batches pulled from the dataset in this RLlib iteration.
+        self.metrics.log_value(
+            DATASET_NUM_ITERS_PER_LEARNER_TRAINED,
+            i,
+            reduce="sum",
+            clear_on_reduce=True,
```
Comment (sven1977) [inline rllib/core/learner/learner.py:1192]: 👍

File: rllib/core/learner/learner_group.py [hunk 10]
```diff
@@ -414,9 +414,15 @@ def _learner_update(
                     " local mode! Try setting `config.num_learners > 0`."
                 )
 
-            if isinstance(batch, list) and isinstance(batch[0], ray.ObjectRef):
+            if isinstance(batch, list):
+                # Ensure we are not in a multi-learner setting.
```
Comment (sven1977) [inline rllib/core/learner/learner_group.py:418]: Cool.

File: rllib/offline/offline_prelearner.py [hunk 11]
```diff
@@ -96,18 +93,12 @@ def __init__(
...
-            self._learner = learner
-            self.learner_is_remote = False
-            self._module = self._learner._module
-        # Otherwise we have remote `Learner`s.
-        else:
-            self.learner_is_remote = True
-            # Build the module from spec. Note, this will be a MultiRLModule.
-            self._module = module_spec.build()
-            self._module.set_state(module_state)
+        ## Build the module from spec. Note, this will be a MultiRLModule.
```
Comment (sven1977) [inline rllib/offline/offline_prelearner.py:96]: nit nit: remove double `#` and just add typehint: `self._module: MultiRLModule = ...` instead of mentioning the type in the comment.

File: rllib/tuned_examples/bc/cartpole_bc.py [hunk 12]
```diff
@@ -51,7 +52,7 @@
         # mode in a single RLlib training iteration. Leave this to `None` to
         # run an entire epoch on the dataset during a single RLlib training
         # iteration. For single-learner mode, 1 is the only option.
-        dataset_num_iters_per_learner=1 if not args.num_learners else None,
+        dataset_num_iters_per_learner=5,
```
Comment (sven1977) [inline rllib/tuned_examples/bc/cartpole_bc.py:55]: I like! More power. :)

File: rllib/tuned_examples/bc/cartpole_bc.py [hunk 13]
```diff
@@ -72,6 +73,14 @@
     )
 )
 
+if not args.no_tune:
+    warnings.warn(
+        "You are running the example with Ray Tune. Offline RL uses "
+        "Ray Data and Ray Data does not interact seamlessly with Ray Tune. "
```
Comment (sven1977) [inline rllib/tuned_examples/bc/cartpole_bc.py:79]:
```suggestion
        "Ray Data, which doesn't interact seamlessly with Ray Tune. "
```

File: rllib/tuned_examples/cql/pendulum_cql.py [hunk 14]
```diff
@@ -81,6 +82,14 @@
     )
 )
 
+if not args.no_tune:
+    warnings.warn(
+        "You are running the example with Ray Tune. Offline RL uses "
+        "Ray Data and Ray Data does not interact seamlessly with Ray Tune. "
```
Comment (sven1977) [inline rllib/tuned_examples/cql/pendulum_cql.py:88]:
```suggestion
        "Ray Data, which doesn't interact seamlessly with Ray Tune. "
```

File: rllib/tuned_examples/marwil/cartpole_marwil.py [hunk 15]
```diff
@@ -71,6 +72,14 @@
     )
 )
 
+if not args.no_tune:
+    warnings.warn(
+        "You are running the example with Ray Tune. Offline RL uses "
+        "Ray Data and Ray Data does not interact seamlessly with Ray Tune. "
```
Comment (sven1977) [inline rllib/tuned_examples/marwil/cartpole_marwil.py:78]:
```suggestion
        "Ray Data, which doesn't interact seamlessly with Ray Tune. "
```

**Review by sven1977 — APPROVED** [review]
Awesome progress with this PR. Thanks so much @simonsays1980 for the hard work. A few suggestions before we can merge.

File: rllib/core/learner/learner.py (hunk 8 above)
Comment (simonsays1980) [reply rllib/core/learner/learner.py:58]: Isn't this how this will be done? These metrics are used inside the `Learner`s and then reported to the main Algorithm that merges the dicts from all Learners. This metric should keep track how many updates a learner indeed took and if there is a divergence between the desired and the taken number of updates.

File: rllib/core/learner/learner.py [hunk 16]
```diff
@@ -1180,6 +1184,18 @@ def _finalize_fn(batch: Dict[str, numpy.ndarray]) -> Dict[str, Any]:
                 value=loss,
                 window=1,
             )
+        # Record the number of batches pulled from the dataset in this RLlib iteration.
+        self.metrics.log_value(
+            DATASET_NUM_ITERS_PER_LEARNER_TRAINED,
```
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/core/learner/learner.py:1189]: Got ya now. Yes this will be aggregated then.

File: rllib/offline/offline_prelearner.py (hunk 11 above)
Comment (simonsays1980) [reply rllib/offline/offline_prelearner.py:96]: Great catch! Makes totally sense.
